typing/ctype.ml

```diff
@@ -1101,6 +1170,38 @@ let compute_univars ty =
   fun ty ->
     try !(TypeHash.find node_univars ty) with Not_found -> TypeSet.empty
 
+(* [type_subexpressions_with_occurrences ids ty]
+    returns the set of type subexpressions of [ty] that contain at least one
+    identifier in [ids].
+*)
+let type_subexpressions_with_occurrences ids ty =
+  let inverted = TypeHash.create 17 in
+  inv_type inverted [] ty;
+  let nodes = ref TypeSet.empty in
+  let rec add_all_parents ids inv =
+    if TypeSet.mem inv.inv_type !nodes
+    then ()
+    else
+      match get_desc inv.inv_type with
+      | Tfunctor (_, id', _, _) ->
+        let id' = Ident.of_unscoped id' in
+        let ids = List.filter (fun id -> not (Ident.same id id')) ids
+        in if ids <> [] then begin
+          nodes := TypeSet.add inv.inv_type !nodes;
+          List.iter (add_all_parents ids) inv.inv_parents
+        end
+      | _ ->
+        nodes := TypeSet.add inv.inv_type !nodes;
+        List.iter (add_all_parents ids) inv.inv_parents
+  in
+  TypeHash.iter (fun ty inv ->
+    match get_desc ty with
+    | Tconstr (p, _, _) | Tobject (_, {contents = Some (p, _)})
+    | Tfunctor (_, _, {pack_path = p}, _) | Tpackage {pack_path = p}
```

Comment:
We also need to list row variable names in `Tvariant _`. Otherwise, after defining
```ocaml
module type T = sig 
  type a = int
 type v = [ `A of a ]
end
module M = struct type a = int type v = [`A of a ] end
let f: (module M:T) -> ([> M.v ] as 'a) -> 'a = fun _ x -> x
let u f  = (f: (module M : T) -> ([> M.v ] as 'a) -> 'a :> (module T) -> _ -> _)
```
the following
```ocaml
let g = u f
``` 
fails with
```
Error: This expression has type (module T) -> ([> `A of M.a ] as 'a) -> 'a
       but an expression was expected of type 'b
```